File: src/ObjectManager/Object.Tes/FilePacks/EsmFile.cs
using OA.Core;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;

namespace OA.Tes.FilePacks
{
    public partial class EsmFile : IDisposable
    {
        const int recordHeaderSizeInBytes = 16;
        public override string ToString() => $"{Path.GetFileName(FilePath)}";
        UnityBinaryReader _r;
        public string FilePath;
        public GameFormatId Format;
        public Dictionary<string, RecordGroup> Groups;

        public EsmFile(string filePath, GameId gameId)
        {
            if (filePath == null)
                return;
            FilePath = filePath;
            Format = GetFormatId();
            _r = new UnityBinaryReader(File.Open(filePath, FileMode.Open, FileAccess.Read));
            var watch = new Stopwatch();
            watch.Start();
            Read(1);
            Utils.Info($"Loading: {watch.Elapsed}");
            Process();
            watch.Stop();
            GameFormatId GetFormatId()
            {
                switch (gameId)
                {
                    // tes
                    case GameId.Morrowind: return GameFormatId.TES3;
                    case GameId.Oblivion: return GameFormatId.TES4;
                    case GameId.Skyrim:
                    case GameId.SkyrimSE:
                    case GameId.SkyrimVR: return GameFormatId.TES5;
                    // fallout
                    case GameId.Fallout3:
                    case GameId.FalloutNV: return GameFormatId.TES4;
                    case GameId.Fallout4:
                    case GameId.Fallout4VR: return GameFormatId.TES5;
                    default: throw new InvalidOperationException();
                }
            }
        }

        void IDisposable.Dispose()
        {
            Close();
        }

        ~EsmFile()
        {
            Close();
        }

        public void Close()
        {
            if (_r != null)
            {
                _r.Close();
                _r = null;
            }
        }

        void Read(int recordLevel)
        {
            var rootHeader = new Header(_r, Format, null);
            if ((Format == GameFormatId.TES3 && rootHeader.Type != "TES3") || (Format != GameFormatId.TES3 && rootHeader.Type != "TES4"))
                throw new FormatException($"{FilePath} record header {rootHeader.Type} is not valid for this {Format}");
            var rootRecord = rootHeader.CreateRecord(rootHeader.Position, recordLevel);
            rootRecord.Read(_r, FilePath, Format);
            // morrowind hack
            if (Format == GameFormatId.TES3)
            {
                var group = new RecordGroup(_r, FilePath, Format, recordLevel);
                group.AddHeader(new Header
                {
                    Label = null,
                    DataSize = (uint)(_r.BaseStream.Length - _r.BaseStream.Position),
                    Position = _r.BaseStream.Position,
                });
                group.Load();
                Groups = group.Records.GroupBy(x => x.Header.Type)
                    .ToDictionary(x => x.Key, x =>
                    {
                        var s = new RecordGroup(_r, FilePath, Format, recordLevel) { Records = x.ToList() };
                        s.AddHeader(new Header { Label = Encoding.ASCII.GetBytes(x.Key) });
                        return s;
                    });
                return;
            }
            // read groups
            Groups = new Dictionary<string, RecordGroup>();
            var endPosition = _r.BaseStream.Length;
            while (_r.BaseStream.Position < endPosition)
            {
                var header = new Header(_r, Format, null);
                if (header.Type != "GRUP")
                    throw new InvalidOperationException($"{header.Type} not GRUP");
                var nextPosition = _r.BaseStream.Position + header.DataSize;
                var label = Encoding.ASCII.GetString(header.Label);
                if (!Groups.TryGetValue(label, out RecordGroup group))
                {
                    group = new RecordGroup(_r, FilePath, Format, recordLevel);
                    Groups.Add(label, group);
                }
                group.AddHeader(header);
                _r.BaseStream.Position = nextPosition;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/EsmFile.swift
//
//  EsmFile.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

import Foundation
import simd

public class EsmFile: CustomStringConvertible {
    static let recordHeaderSizeInBytes = 16
    public var description: String { return "\(filePath)" }
    var _r: BinaryReader!
    public let filePath: String
    public let format: GameFormatId
    public var groups: [String : RecordGroup]? = nil
    // TES3
    var _MANYsById: [String : Record]!
    var _LTEXsById: [Int64 : LTEXRecord]!
    var _LANDsById: [int3 : LANDRecord]!
    var _CELLsById: [int3 : CELLRecord]!
    var _CELLsByName: [String : CELLRecord]!
    // TES4
    var _WRLDsById: [UInt32 : (x: WRLDRecord, y: [RecordGroup])]!
    var _LTEXsByEid: [String : LTEXRecord]!

    init(_ filePath: URL?, for game: GameId) {
        func getFormatId() -> GameFormatId {
            switch game {
            // tes
            case .morrowind: return .TES3
            case .oblivion: return .TES4
            case .skyrim, .skyrimSE, .skyrimVR: return .TES5
            // fallout
            case .fallout3, .falloutNV: return .TES4
            case .fallout4, .fallout4VR: return .TES5
            }
        }
        self.format = getFormatId()
        guard let filePath = filePath else {
            self.filePath = "missing"
            return
        }
        self.filePath = filePath.path
        _r = BinaryReader(FileBaseStream(path: self.filePath)!)
        let start = Date()
        read(recordLevel: 1)
        let endRead = Date()
        debugPrint("Loading: \(endRead.timeIntervalSince(start))")
        process()
    }

    deinit {
        close()
    }

    public func close() {
        _r?.close()
        _r = nil
    }

    func read(recordLevel: Int) {
        let rootHeader = Header(_r, for: format, parent: nil)
        guard (format != .TES3 || rootHeader.type == "TES3") && (format == .TES3 || rootHeader.type == "TES4") else {
            fatalError("\(filePath) record header \(rootHeader.type) is not valid for this \(format)")
        }
        let rootRecord = rootHeader.createRecord(at: rootHeader.position, recordLevel: recordLevel)!
        rootRecord.read(_r, filePath, for: format)
        // morrowind hack
        guard format != .TES3 else {
            let group = RecordGroup(_r, filePath, for: format, recordLevel: recordLevel)
            group.addHeader(Header(
                label: nil,
                dataSize: UInt32(_r.baseStream.length - _r.baseStream.position),
                position: _r.baseStream.position
            ));
            group.load()
            groups = Dictionary(uniqueKeysWithValues:
                Dictionary(grouping: group.records) { $0.header.type }.map { x -> (key: String, value: RecordGroup) in
                    (key: x.key, value: RecordGroup(_r, filePath, for: format, recordLevel: recordLevel, label: x.key, records: x.value)) })
            return
        }
        // read groups
        groups = [String : RecordGroup]()
        let endPosition = _r.baseStream.length
        while _r.baseStream.position < endPosition {
            let header = Header(_r, for: format, parent: nil)
            guard header.type == "GRUP" else {
                fatalError("\(header.type) not GRUP")
            }
            let nextPosition = _r.baseStream.position + UInt64(header.dataSize)
            let label = String(data: header.label!, encoding: .ascii)!
            var group = groups![label]
            if group == nil {
                group = RecordGroup(_r, filePath, for: format, recordLevel: recordLevel)
                groups![label] = group
            }
            group!.addHeader(header)
            _r.baseStream.position = nextPosition
        }
    }
}
